autoprecompiles/src/range_constraint_optimizer.rs

```diff
@@ -0,0 +1,167 @@
+                    let existing_rc = range_constraints
+                        .entry(expr)
+                        .or_insert_with(RangeConstraint::default);
+                    *existing_rc = existing_rc.conjunction(&rc);
+                }
+                false
+            }
+            None => true,
+        }
+    });
+
+    // Filter range constraints that are already implied by existing constraints.
+    // TODO: They could also be implied by each other.
+    let mut solver = new_solver(system.clone(), bus_interaction_handler.clone());
+    solver.solve().unwrap();
+    let to_constrain = to_constrain
+        .into_iter()
+        .unique()
+        .map(|expr| {
+            let rc = range_constraints.remove(&expr).unwrap();
+            (expr, rc)
+        })
+        .filter(|(expr, rc)| {
+            let current_rc = solver.range_constraint_for_expression(expr);
+            current_rc != current_rc.conjunction(rc)
+        })
+        .collect::<Vec<_>>();
+
+    // Implement bit constraints via polynomial constraints, if the degree bound allows.
+    let mut bit_constraints = Vec::new();
+    let to_constrain = to_constrain
+        .into_iter()
+        .filter(|(expr, rc)| {
+            if rc == &RangeConstraint::from_mask(1) && expr.degree() < degree_bound.identities {
+                bit_constraints.push(expr.clone() * (expr.clone() - GroupedExpression::one()));
+                false
+            } else {
+                true
+            }
+        })
+        .collect();
+
+    // Create all range constraints in batch and add them to the system.
+    let range_constraints = bus_interaction_handler.batch_make_range_constraints(to_constrain);
+    for bus_interaction in &range_constraints {
+        assert_eq!(bus_interaction.multiplicity.try_to_number(), Some(T::one()));
+    }
+    system.bus_interactions.extend(range_constraints);
+    system.algebraic_constraints.extend(bit_constraints);
+
+    system
+}
+
+/// Utility functions useful for implementing `batch_make_range_constraints`.
+pub mod utils {
+    use itertools::Itertools;
+    use powdr_constraint_solver::{
+        grouped_expression::GroupedExpression, range_constraint::RangeConstraint,
+    };
+    use powdr_number::FieldElement;
+    use std::fmt::Display;
+    use std::hash::Hash;
+
+    use crate::range_constraint_optimizer::RangeConstraints;
+
+    /// If the range constraints is the range 0..(2^bits - 1), returns Some(bits).
+    pub fn range_constraint_to_num_bits<T: FieldElement>(
+        range_constraint: &RangeConstraint<T>,
+    ) -> Option<usize> {
+        (0..30).find(|num_bits| {
+            let mask = (1u64 << num_bits) - 1;
+            range_constraint == &RangeConstraint::from_mask(mask)
+        })
+    }
+
+    /// Given a a set of range constraints, filters out those which can be checked via a
+    /// byte constraint.
+    pub fn filter_byte_constraints<T: FieldElement, V: Ord + Clone + Eq + Display + Hash>(
```

Comment:
Can you document what this function returns?
Does it return an expression `x` such that `x in [0..=265]` is equivalent to the input?
Is that really correct? If the input is a 0/1 constraint on the expr `x` then it would return `128 * x`. There should be plenty of other values to fulfill that constraint, for example if `x` is the inverse of 128, then we would get `1`, which is less than 256